File: pset6/readability/readability.py
from cs50 import get_string
import re


def main():

    text = get_string("Text: ")
    details = detail(text)
    calculation(details)


def detail(text):
    details = [0, 1, 0]
    for char in text:
        if char.isalpha():                # If the current character is a letter it adds to the number of letters of the text
            details[0] += 1

        elif char.isspace():             # If the current character is a space it adds to the number of words for the text
            details[1] += 1

        elif char in ["?", ".", "!"]:     # If the current character is punctuation indicating the end of a sentence, the number of sentences
            details[2] += 1                     # of the text is increased by one

    return details


def calculation(details):

    letters = details[0]                        # Initializes letters with its corespondant in details
    words = details[1]                          # Initializes words with its correspondant in details
    sentences = details[2]                      # Initializes sentences with its correspondant in details

    L = (letters / words) * 100                 # Finds the number of letters per 100 words in the text
    S = (sentences / words) * 100               # Finds the number of sentences per 100 words in the text

    grade = 0.0588 * L - 0.296 * S - 15.8       # Performs the Coleman-Liau index
    grade = round(grade)

    if grade < 1:
        print("Before Grade 1")

    elif grade >= 16:
        print("Grade 16+")

    else:
        print(f"Grade {grade}")


main()

File: pset2/readability/readability.c
//Header files needed
#include <stdio.h>
#include <math.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h> 

//Prototypes
int num_of_words(string s);
int num_of_letters(string s);
int num_of_sentences(string s);
void reading_level(int x, int y, int z);

int main(void)
{
    //Getting input of text, which will be checked for its reading level, from user
    string text = get_string("Text: ");
    
    //Initializing variables need for the Coleman-Liau index
    int letters = num_of_letters(text);
    int words = num_of_words(text);
    int sentences = num_of_sentences(text);
    
    //Calculation of reading level and printing it to screen
    reading_level(letters, words, sentences);
    
    //Shows sucessful exit
    return 0;
}

int num_of_letters(string s)
{
    int letters = 0;
    
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        //Checks if character is a letter
        if (isalpha(s[i]))
        {
            letters++;
        }
    }
    
    return letters; 
}

int num_of_words(string s)
{
    int words = 1;
    
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        //Checks if character is a whitespace in order to determine a word
        if (isspace(s[i]))
        {
            words++;
        }
    }
    
    return words; 
}

int num_of_sentences(string s)
{
    int sentences = 0;
    
    for (int i = 0, n = strlen(s); i < n; i++)
    {
        //Checks if character is '.', '?' or '!'
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            sentences++;
        }
    }
    
    return sentences; 
}

void reading_level(int x, int y, int z)
{
    //Coleman-Liau index
    double L = (x / (float) y) * 100;
    double S = (z / (float) y) * 100;
    double index = 0.0588 * L - 0.296 * S - 15.8;
    
    int level = round(index);
    
    //Checks the range the level lies in
    if (level > 1 && level < 16)
    {
        printf("Grade %i\n", level);
    }
    
    else if (level >= 16)
    {
        printf("Grade 16+\n");
    }
    
    else  
    {
        printf("Before Grade 1\n");
    }
}